sdk/macro-frozen-abi/src/lib.rs

```diff
@@ -0,0 +1,461 @@
+                let sample_struct = <#type_name #ty_generics>::sample();
+                Serialize::serialize(&sample_struct, digester.create_new())
+            }
+        }
+    }).into()
+}
+
+#[cfg(RUSTC_WITH_SPECIALIZATION)]
+fn derive_abi_digest_enum_type(input: ItemEnum) -> TokenStream {
+    let type_name = &input.ident;
+    let mut serialized_variants = quote! {};
+    let mut variant_count = 0;
+    let (impl_generics, ty_generics, where_clause) = input.generics.split_for_impl();
+    for variant in &input.variants {
+        // Don't digest a variant with serde(skip)
+        if filter_serde_attrs(&variant.attrs) {
+            continue;
+        };
+        let sample_variant = quote_sample_variant(&type_name, &ty_generics, &variant);
+        variant_count += 1;
+        serialized_variants.extend(quote! {
+            #sample_variant;
+            Serialize::serialize(&sample_variant, digester.create_enum_child())?;
+        });
+    }
+
+    let type_str = format!("{}", type_name);
+    (quote! {
+        impl #impl_generics ::solana_sdk::abi_digester::AbiDigest for #type_name #ty_generics #where_clause {
+            fn abi_digest(digester: &mut ::solana_sdk::abi_digester::AbiDigester) -> ::solana_sdk::abi_digester::DigestResult {
+                let enum_name = #type_str;
+                use ::serde::ser::Serialize;
+                use ::solana_sdk::abi_digester::AbiSample;
+                digester.update_with_string(format!("enum {} (variants = {})", enum_name, #variant_count));
+                #serialized_variants
+                Ok(digester.create_child())
+            }
+        }
+    }).into()
+}
+
+#[cfg(RUSTC_WITH_SPECIALIZATION)]
+#[proc_macro_derive(AbiDigest)]
+pub fn derive_abi_digest(item: TokenStream) -> TokenStream {
+    let item = parse_macro_input!(item as Item);
+
+    match item {
+        Item::Struct(input) => derive_abi_digest_struct_type(input),
+        Item::Enum(input) => derive_abi_digest_enum_type(input),
+        _ => panic!("not applicable to {:#?}", item),
+    }
+}
+
+#[cfg(RUSTC_WITH_SPECIALIZATION)]
+fn quote_for_test(
+    test_mod_ident: &Ident,
+    type_name: &Ident,
+    expected_digest: &str,
+) -> TokenStream2 {
+    // escape from nits.sh...
+    let p = Ident::new(&("ep".to_owned() + "rintln"), Span::call_site());
+    quote! {
+        #[cfg(test)]
+        mod #test_mod_ident {
+            use super::*;
+            use ::solana_sdk::abi_digester::AbiDigest;
+
+            #[test]
+            fn test_digest() {
+                ::solana_logger::setup();
+                let mut digester = ::solana_sdk::abi_digester::AbiDigester::create();
+                let r = <#type_name>::abi_digest(&mut digester);
+                let mut hash = digester.finalize();
+                if r.is_err() {
+                    ::log::error!("digest error: {:#?}", r);
+                }
+                r.unwrap();
+                if let Ok(_) = ::std::env::var("SOLANA_ABI_BULK_UPDATE") {
```

Comment:
my most favorite part of code in this huge PR for its productivity improvement. ;)